src/main/java/jenkins/branch/OrganizationFolder.java

```diff
@@ -0,0 +1,247 @@
+                    return listener;
+                }
+                @Override public SCMSourceObserver.ProjectObserver observe(final String projectName) {
+                    if (projects.containsKey(projectName)) {
+                        throw new IllegalArgumentException("Will not create duplicate subproject named " + projectName);
+                    }
+                    return new ProjectObserver() {
+                        List<SCMSource> sources = new ArrayList<SCMSource>();
+                        @Override public void addSource(SCMSource source) {
+                            sources.add(source);
+                            source.setOwner(OrganizationFolder.this);
+                        }
+                        @Override public void addAttribute(String key, Object value) throws IllegalArgumentException, ClassCastException {
+                            throw new IllegalArgumentException();
+                        }
+                        @Override public void complete() throws IllegalStateException, InterruptedException {
+                            if (sources == null) {
+                                throw new IllegalStateException();
+                            }
+                            for (MultiBranchProjectFactory factory : projectFactories) {
+                                MultiBranchProject<?, ?> project;
+                                try {
+                                    project = factory.createProject(OrganizationFolder.this, projectName, sources, Collections.<String,Object>emptyMap(), listener);
+                                } catch (InterruptedException x) {
+                                    throw x;
+                                } catch (Exception x) {
+                                    x.printStackTrace(listener.error("Failed to create a subproject " + projectName));
+                                    continue;
+                                }
+                                if (project != null) {
+                                    List<BranchSource> branchSources = project.getSourcesList();
+                                    for (SCMSource source : sources) {
+                                        // TODO we probably need to define a BranchPropertyFactory
+                                        branchSources.add(new BranchSource(source, new DefaultBranchPropertyStrategy(new BranchProperty[0])));
+                                    }
+                                    projects.put(projectName, project);
+                                    break;
+                                }
+                            }
+                        }
+                    };
+                }
+                @Override public void addAttribute(String key, Object value) throws IllegalArgumentException, ClassCastException {
+                    throw new IllegalArgumentException();
+                }
+            });
+        }
+        return projects;
+    }
```

Comment:
Mm, not sure. It actually does require some thought to see what happens in what order, and the current source layout is probably as close to intuitive as you are going to get. In fact I used `SCMSourceObserver` as an opportunity to _change_ the order: now `MultiBranchProjectFactory.createProject` etc. happens as soon as `complete` is called, before other repositories have been discovered; the previous API forced `computeChildren` to wait until the navigator was done before it took any action. (Currently `ComputedFolder` does not go ahead and add the child and call `onCreatedFromScratch` and `afterNewItemCreated` until the folder recomputation is done, though I think it could be improved to be more incremental. Deleting orphaned children does have to wait until the end.)